components/odh-notebook-controller/controllers/notebook_rbac.go

```diff
@@ -0,0 +1,166 @@
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     roleRefKind,
+			Name:     roleRefName,
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+}
+
+// checkRoleExists checks if a Role or ClusterRole exists in the namespace.
+func (r *OpenshiftNotebookReconciler) checkRoleExists(ctx context.Context, roleRefKind, roleRefName, namespace string) (bool, error) {
+	if roleRefKind == "ClusterRole" {
+		// Check ClusterRole if roleRefKind is ClusterRole
+		clusterRole := &rbacv1.ClusterRole{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName}, clusterRole)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// ClusterRole not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	} else {
+		// Check Role if roleRefKind is Role
+		role := &rbacv1.Role{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName, Namespace: namespace}, role)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// Role not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	}
+	return true, nil // Role or ClusterRole exists
+}
+
+// Helper function to delete the RoleBinding if the notebooks deleted
+func (r *OpenshiftNotebookReconciler) deleteRoleBinding(ctx context.Context, rolebindingName, namespace string) error {
+	roleBinding := &rbacv1.RoleBinding{}
+	err := r.Get(ctx, types.NamespacedName{Name: rolebindingName, Namespace: namespace}, roleBinding)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			return nil // RoleBinding not found, nothing to delete
+		}
+		return err // Some other error occurred
+	}
+
+	// Delete the RoleBinding
+	return r.Delete(ctx, roleBinding)
+}
+
+// reconcileRoleBinding manages creation, update, and deletion of RoleBindings and ClusterRoleBindings
+func (r *OpenshiftNotebookReconciler) reconcileRoleBinding(
+	notebook *nbv1.Notebook, ctx context.Context, rolebindingName, roleRefKind, roleRefName string) error {
+
+	log := r.Log.WithValues("notebook", types.NamespacedName{Name: notebook.Name, Namespace: notebook.Namespace})
+
+	// Check if the Role or ClusterRole exists before proceeding
+	roleExists, err := r.checkRoleExists(ctx, roleRefKind, roleRefName, notebook.Namespace)
+	if err != nil {
+		log.Error(err, "Error checking if Role exists", "Role.Kind", roleRefKind, "Role.Name", roleRefName)
+		return err
+	}
+	if !roleExists {
+		return nil // Skip if dspa Role is not found on the namespace
+	}
+
+	// Create a new RoleBinding based on provided parameters
+	roleBinding := NewRoleBinding(notebook, rolebindingName, roleRefKind, roleRefName)
+
+	// Check if the RoleBinding already exists
+	found := &rbacv1.RoleBinding{}
+	err = r.Get(ctx, types.NamespacedName{Name: rolebindingName, Namespace: notebook.Namespace}, found)
+	if err != nil && apierrs.IsNotFound(err) {
+		log.Info("Creating RoleBinding", "RoleBinding.Namespace", roleBinding.Namespace, "RoleBinding.Name", roleBinding.Name)
+		err = r.Create(ctx, roleBinding)
```

Comment:
can you add the setcontrollerreference for setting the ownerreference
example: https://github.com/opendatahub-io/kubeflow/blob/9d606be9d421bcfc388294c7a018de66016ffd8e/components/odh-notebook-controller/controllers/notebook_network.go#L82

when tried on the cluster, i m not noticing ownerreference on rbac

Reply:
I thought with this https://github.com/opendatahub-io/kubeflow/pull/434#discussion_r1822536888 was enough in regards the owners. However, only with that the deletion didn't take action by itself and thats why i added the deleteRbac helper function. 

Now, using the `SetControllerReference` we have enriched the metadata, and the deletion is working without the helper delete function. 

Now the look like:
```
  ownerReferences:
    - apiVersion: kubeflow.org/v1
      kind: Notebook
      name: f1
      uid: 88b9ff85-c6da-479f-9424-bd015550a163
      controller: true
      blockOwnerDeletion: true
```

before
```
  ownerReferences:
    - apiVersion: kubeflow.org/v1beta1
      kind: Notebook
      name: q1
      uid: 84fb0090-0258-48f7-86ff-6fb7d7272075
```